src/Servers/IIS/IISIntegration/src/IISMiddleware.cs

```diff
@@ -147,5 +156,39 @@ public async Task Invoke(HttpContext httpContext)
 
             await _next(httpContext);
         }
+
+        private WindowsPrincipal GetUser(HttpContext context)
+        {
+            var tokenHeader = context.Request.Headers[MSAspNetCoreWinAuthToken];
+
+            if (!StringValues.IsNullOrEmpty(tokenHeader)
+                && int.TryParse(tokenHeader, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var hexHandle))
+            {
+                // Always create the identity if the handle exists, we need to dispose it so it does not leak.
+                var handle = new IntPtr(hexHandle);
+                var winIdentity = new WindowsIdentity(handle, IISDefaults.AuthenticationScheme);
+
+                // WindowsIdentity just duplicated the handle so we need to close the original.
+                NativeMethods.CloseHandle(handle);
```

Comment:
I know it's not new, but should this be in a finally?

Reply:
The most likely thing to throw would be for an invalid handle, so closing it wouldn't help much. As this dates back to 1.0 and we haven't seen any issues, I'm going to leave it 😁.